### PR TITLE
MB-10592 Use custom nullable types for patch requests ADR

### PR DESCRIPTION
**Pull request by eamahanna**
## Summary

This ADR discusses how to handle setting fields to null via update requests. Link at at the bottom of the adr for additional information.

A prototype PR is here: https://github.com/transcom/mymove/pull/7881

### CODE REVIEW

File: docs/adr/0066-use-custom-nullable-types-for-patch-requests.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Use custom nullable types for patch requests
+
+**User Story:** [MB-10592](https://dp3.atlassian.net/browse/MB-10592)
+
+There are several instances within the milmove app where a user will need to delete a field or reset a field to null on an object. Currently using patch requests this is not possible because there is no way to differentiate between a field that was omitted or a field that is intentially set to null in golang.
```
Comment (leia-sefkin) [inline docs/adr/0066-use-custom-nullable-types-for-patch-requests.md:5]: Edit to add 'no way to differentiate between a *JSON* field that was omitted...'

File: docs/adr/0066-use-custom-nullable-types-for-patch-requests.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+## Considered Alternatives
+
+* Use custom nullable types
+* Use PUT requests in addition to PATCH requests
+
+## Decision Outcome
+
+* Chosen Alternative: Use custom nullable types
+To solve this probably engineers will now be able to use a custom, nullable type that indicates if the field was set to null or not. This solution was selected because the implementation of creating the nullable types is realtively quick, and it allows milmove to work with a single update request type: PATCH. Using the custom types will require engineers to alter how they access the value of the nullable fields, and have additional understand of how this type works. The maintenance of the custom code should be minimal after implementation.
```
Comment (leia-sefkin) [inline docs/adr/0066-use-custom-nullable-types-for-patch-requests.md:15]: To solve this ~probably~ engineers...

I might also edit the end of the first sentence to clarify what the nullable type means - that we want to clear the data. Something like '...nullable type that indicates if the field was explicitly set to null [delete the data] vs implicitly [don't change anything]...'

**Review by ahobson — APPROVED** [review]
Thank you! I think this captures the situation quite well

**Review by leia-sefkin — APPROVED** [review]
Added a couple small comments for extra clarity but overall LGTM

**Review by reggieriser — COMMENTED** [review]
This does a good job representing the problem and alternatives.  I added a few notes about some potential clarifications and nits for your consideration.

File: docs/adr/0066-use-custom-nullable-types-for-patch-requests.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+* `+` Allows milmove to use a single update request type
+* `+` Allows API users to set fields to null without having to send all object data
+* `-` Custom code will require milmove engineers to be aware of when and why to use nullable types. There may be a small learning curve.
+
+### Use PUT requests in addition to PATCH requests
+
+* `+` PUT requests are a widely understood and known method of making updates
+* `+` Does not require custom code to implement
+* `-` Requires the API user to return all object data when updating a single field
+* `-` Will require adding additional support for another endpoint for a given resource (new handler, service, changes to yaml, FE calls/payloads, ect.)
```
Comment (reggieriser) [inline docs/adr/0066-use-custom-nullable-types-for-patch-requests.md:32]: Nit: `ect.` -> `etc.`

**Review by reggieriser — APPROVED** [review]
Changes LGTM!  Thanks for making them.  Noted one small typo in the changes.

**Pull request merged**